presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestPushProjectionIntoTableScan.java

```diff
@@ -0,0 +1,210 @@
+import io.prestosql.sql.tree.Identifier;
+import io.prestosql.sql.tree.SymbolReference;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.RowType.field;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.expression;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.project;
+import static io.prestosql.sql.planner.assertions.PlanMatchPattern.tableScanWithProvidedHandles;
+import static io.prestosql.testing.TestingSession.testSessionBuilder;
+import static java.util.Arrays.asList;
+import static org.testng.Assert.assertEquals;
+
+@Test
+public class TestPushProjectionIntoTableScan
+{
+    private static final String MOCK_CATALOG = "mock_catalog";
+    private static final String TEST_SCHEMA = "test_schema";
+    private static final String TEST_TABLE = "test_table";
+    private static final SchemaTableName TEST_SCHEMA_TABLE = new SchemaTableName(TEST_SCHEMA, TEST_TABLE);
+    private static final Type ROW_TYPE = RowType.from(asList(field("a", BIGINT), field("b", BIGINT)));
+
+    private static final TableHandle TEST_TABLE_HANDLE = createTableHandleForMockConnector(TEST_SCHEMA, TEST_TABLE);
+
+    private static final Session MOCK_SESSION = testSessionBuilder().setCatalog(MOCK_CATALOG).setSchema(TEST_SCHEMA).build();
+
+    public TestPushProjectionIntoTableScan() {}
+
+    @Test
+    public void testDoesNotFire()
+    {
+        RuleTester ruleTester = new RuleTester();
+
+        try {
+            String inputColumnName = "input_column";
+            Type inputColumnType = ROW_TYPE;
+            ColumnHandle inputColumnHandle = new TpchColumnHandle(inputColumnName, inputColumnType);
+
+            MockConnectorFactory factory = createMockFactoryWithoutApplyProjection(ImmutableMap.of(inputColumnName, inputColumnHandle));
+
+            ruleTester.createCatalog(MOCK_CATALOG, factory, ImmutableMap.of());
+
+            ruleTester.assertThat(new PushProjectionIntoTableScan(ruleTester.getMetadata(), new TypeAnalyzer(new SqlParser(), ruleTester.getMetadata())))
+                    .on(p -> {
+                        Symbol symbol = p.symbol(inputColumnName, inputColumnType);
+                        return p.project(
+                                Assignments.of(p.symbol("expr_deref", BIGINT), new DereferenceExpression(symbol.toSymbolReference(), new Identifier("a"))),
+                                p.tableScan(TEST_TABLE_HANDLE, ImmutableList.of(symbol), ImmutableMap.of(symbol, inputColumnHandle)));
+                    })
+                    .withSession(MOCK_SESSION)
+                    .doesNotFire();
+        }
+        finally {
+            ruleTester.close();
+        }
+    }
+
+    @Test
+    public void testPushProjection()
+    {
+        RuleTester ruleTester = new RuleTester();
```

Comment:
Can we instantiate `RuleTester` once only?

Reply:
I see that instantiating in once will avoid the overhead of re-creating the query runner.  

In that case, we'll need to either 

(1) remove mock catalog and connector factory in every testcase: 
- ConnectorManager doesn't provide methods to remove a connector factory. Is this by design?

(2) Create a mock catalog and a factory with different names every time:
- This would require changing MockConnectorFactory to have different names, which is different from the pattern we follow for connector factory names. But I think it should be okay to do so.

@findepi What is your preference here?

Reply:
How lonsg does it take to instantiate the `RuleTester`?

Reply:
@findepi Benchmarked the following snippet through JMX, which came out to be  11.585 ± 1.334 ops/s. roughly translates to 86ms per iteration.

```
        RuleTester ruleTester = new RuleTester();
        return ruleTester;
```